src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -223,7 +224,10 @@
 
         return CtpQueryUtils
             .queryAll(syncOptions.getCtpClient(), keysQueryMapper.apply(keysNotCached), pageConsumer)
-            .thenApply(result -> keyToIdCache.asMap());
+            .thenApply(result -> {
+                keyToIdCache.cleanUp();
```

Comment:
Why we need a cleanup, and what is the benefit? 

I would assume the cleanup will be run when the cache reached some limit? Did you try it out?

Reply:
OK, I see what you did here, I tested quickly like below and understand eviction is not automatic.

```` java
     final Cache<String, String> testCache = Caffeine.newBuilder()
                                                    .maximumSize(10)
                                                    .build();
        for (int i = 0; i < 100; i++) {
            testCache.put("key" + i, "val" + i);
        }
````

Reply:
Yes, cleanup is running when cache reached the size limit asynchronously. Actually i added to be able to test the eviction. Cannot access the `keyToIdCache` field from tests... - i could make it accessible for the test. WDYT?

Reply:
But the problem is different because the code is using locks/threads and not really performant to run for every call 🤔

Reply:
I wrote a simple method like this 
```
        Cache<String, String> cache = Caffeine.newBuilder().maximumSize(10).build();

        for (int i = 0; i<100 ; ++i) {
            cache.put(String.valueOf(i), String.valueOf(i));

        }
        System.out.println(cache.estimatedSize()); // printout 100
        Thread.sleep(1000);
        System.out.println(cache.estimatedSize()); // printout 10
```
As mentioned, cleanup is running when cache reached the size limit asynchronously. It automatically triggers and takes a second to do eviction. So I guess we don't need to `cleanUp()` manually.

Reply:
Another small test
 ```
       Cache<String, String> cache = Caffeine.newBuilder()
                                                .maximumSize(10)
                                                .build();

        for (int i = 0; i<100 ; ++i) {
            Thread.sleep(10);
            cache.put(String.valueOf(i), String.valueOf(i));

        }
        System.out.println(cache.estimatedSize()); // printout 11
        Thread.sleep(1000);
        System.out.println(cache.estimatedSize()); // printout 10
```
It proves that high cache size produced before because cache input is quicker than eviction.

Reply:
@salander85 I don't suggest to change method access modifier easily for test case purpose. Your added unit test case is good enough to verify eviction. I can't find any other scenario we can verify the eviction.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/628/commits/120f43bb2d59c9bbc3f856c86d3a85bba651f21c

Reply:
However, without explicitly cleanup the cache in every call, i found the test class fails all the time. I have to insert few milliseconds delay in eviction test case to make it run properly. I guess cache handles the operation asynchronously when running test class and cause the problem.

Reply:
fyi, you can use `Caffeine.executor(Runnable::run)` to perform the eviction on the calling thread.

Reply:
67aad19

Reply:
@ben-manes thanks for the suggestion 👍🏽